ENH: don't coerce input to double with custom metric in cdist and pdist

apbard opened the pull request:
When using pdist with a custom metric there is no need to coerce input to be of type double. this will allow to use for instance an edit distance metric.

person142 commented:
Should also add a test--pass in a custom metric not expecting double inputs and make sure it works.

apbard commented:
ok, i'll do it. Should I wrote a "true" metric to be added to scipy or a simple distance metric as a closure in test would be enough?

person142 commented:
As a closure in a test is fine. (Probably preferable because it involves much less work.)

person142 commented:
Also note there were a couple pep8 issues:

https://travis-ci.org/scipy/scipy/jobs/165549010

They're easily fixed though.

apbard commented:
I have added the test for a generic function that expects a custom object and done the same thing also for cdist.

person142 commented:
Sorry for taking so long to get back to this! I think the changes here are reasonable, but I'm not super familiar with `spatial` so I will wait for someone else to take a look.

rgommers commented on scipy/spatial/distance.py:
```diff
@@ -1217,20 +1217,22 @@ def pdist(X, metric='euclidean', p=2, w=None, V=None, VI=None):
         if metric == minkowski:
             def dfun(u, v):
                 return minkowski(u, v, p)
+            X = _convert_to_double(X)
```
Either converting to double is needed for all built-in distance functions or none of them I'd think. Since most are going through the ``else`` clause here, this looks inconsistent.

apbard replied:
If you remove the conversion some tests are failing

rgommers replied:
Hmm, in that case I expect that other built in metrics will also need this conversion.

rgommers commented on scipy/spatial/distance.py:
```diff
@@ -1994,8 +1996,8 @@ def cdist(XA, XB, metric='euclidean', p=2, V=None, VI=None, w=None):
 
     # The C code doesn't do striding.
-    XA = _copy_array_if_base_present(_convert_to_double(XA))
-    XB = _copy_array_if_base_present(_convert_to_double(XB))
```
It also seems inconsistent to completely remove the conversion from ``cdist``, but leave it in for ``pdist``.

rgommers replied:
Looks like there was duplicate conversion - all calls to ``_distance_wrap.cdist_*`` still have the correct conversions further down in this function.

rgommers commented:
The ``seuclidean`` and ``mahalanobis`` clauses aren't tested (https://codecov.io/gh/scipy/scipy/compare/cac9bca6d434c6b7ce6b03df3ff5a53a044d7c2d...202123db443844522f85cef3a89e598007923d60), and the test coverage for the ``else`` clause probably isn't complete. Would be good to have a single test like:

    for func in [hamming, jaccard, seuclidean, ...]:  # all distance metrics
        pdist(X, func)

apbard commented:
I have noticed now that those clauses aren't tested at all:
https://codecov.io/gh/scipy/scipy/src/ef933c31cc9effd9731dbb8cbbe6a8a06eb3c6bc/scipy/spatial/distance.py

I agree with you that more tests are needed but maybe they deserve a standalone PR

rgommers commented:
Overall I think this should be fine. To do is removing the 4 casts to double in ``pdist`` and fixing the test coverage.

rgommers commented:
> I agree with you that more tests are needed but maybe they deserve a standalone PR

The test coverage for those two ``elif``s is a minor thing, will take 5 min to fix. Testing with ``int`` and ``float32`` is more work, and that definitely belongs in this PR.

apbard commented:
perfect, I'll try to do it

apbard commented:
I have implemented tests to check that the different ways to call pdist/cdist are equivalent and to check that input with not double type are cast to the type required by the built-in metrics.
I have also corrected a couple of wrong tests and reduced the duplicated code in both functions.

rgommers commented:
LGTM now, merged. Thanks @apbard, nice tests that you added. Sorry it took a while to get back to this.

metaperture commented:
I'm having issues merging this PR with https://github.com/scipy/scipy/pull/6931 -- it looks like this PR stripped out significant functionality regarding automatically calculating V and VI based weights when they weren't given. I can probably make this tests work again if I take out that functionality like this PR did, but I didn't see a discussion here about removing that functionality, so I wanted to check.

Specifically, I'm talking about the large code blocks removed in https://github.com/scipy/scipy/commit/13df2821. This would definitely break backwards compatibility for anyone who was using cdist to automatically calculate variance.

metaperture commented:
When I was working on tests, I noticed https://github.com/scipy/scipy/issues/1403 -- I've changed behavior so that now completely dissimilar vectors get `inf` distance in yule, dice, and sokalsneath. Previously, sokalsneath would raise a ValueError and yule and dice would raise ZeroDivisionError's.

rgommers commented:
@apbard I just noticed that ``test_cdist_calling_conventions`` takes about half a minute. Would you be able to have a look at that? 

Ideally cut down the time, or otherwise mark that test with ``@dec.slow`` (those run only with ``test('full')`` and if needed add back a small subset to still have decent test coverage with plan ``test()``.

apbard commented:
@rgommers I noticed that too. That's why I have reduced the dataset size. I'll try to reduce it further. 
What is the maximum running time we can afford?
Anyway is there a way to understand if we are on 'full' mode or 'fast' one?

rgommers commented:
> What is the maximum running time we can afford?

For fast we're trying to stay below 1-2 sec per test, and a couple of minutes for all of ``scipy.test()``. For the full test suite up to ~30 sec is acceptable, but of course less is better. We often have trouble with the full test suite exceeding 50 min on TravisCI's underpowered instances.

> Anyway is there a way to understand if we are on 'full' mode or 'fast' one?

The TravisCI tests run the full test suite once (https://github.com/scipy/scipy/blob/master/.travis.yml#L62) and the other Python/Numpy versions are tested with fast.

apbard commented:
ok, I'll look into it. 

> The TravisCI tests run the full test suite once (https://github.com/scipy/scipy/blob/master/.travis.yml#L62) and the other Python/Numpy versions are tested with fast.

But within the test we can't  know if we are in the first case or not, can we?

rgommers commented:
You can - if there's no ``@dec.slow`` decorator (from ``np.testing``) for a test then it always gets run, if there is then it only gets run for ``'full'``.

apbard commented:
Sorry, I am not explaining myself...I meant, can we understand it while running the test?
i.e., doing something like

    if fast:
        do quick checks
    else:
        do full tests

rgommers commented:
Ah no, I don't think that's possible.
